### DISCUSSION TITLE
Enhancing Layout Control and Standard HTML/CSS Compatibility in NiceGUI

### DISCUSSION THREAD

**Discussion by Xtreemrus** [zauberzeug/nicegui#2293] [Ideas / Feature Requests]
Hello NiceGUI Community,

I've been exploring the capabilities of NiceGUI and appreciate its simplicity and the powerful features it offers. However, I've noticed a challenge that I believe could be addressed to make NiceGUI even more appealing, especially to developers who are familiar with standard HTML/CSS frameworks like Tailwind CSS.

**Main Issue**:
The current layout behavior in NiceGUI seems to deviate from the standard behavior expected by developers who are accustomed to traditional HTML/CSS. This makes it difficult for someone following tutorials on Tailwind or CSS to achieve the desired page layout in NiceGUI.

For example, consider the following Tailwind CSS example where three blockquotes are displayed inline, expanding in height as needed: [TailwindCSS Playground Link](https://play.tailwindcss.com/acMVfuSzYE).

In contrast, when attempting to replicate this in NiceGUI, the elements behave differently, often wrapping to the next line and attempting to maintain a consistent height first and foremost.

**Suggestion**:
I propose that we explore ways to allow for more standard HTML layout behavior and Tailwind CSS compatibility with minimal framework interference in the styling. This would empower developers who need to create landing pages/sites with the storage and bindings provided by NiceGUI but do not wish to delve into JavaScript.

Here's a basic Python example for NiceGUI that tries to mimic the Tailwind CSS behavior:

```python
from nicegui import ui

@ui.page('/')
def main():
    ui.add_head_html('''
     <style>
        :root {
            --nicegui-default-padding: 0rem;
            --nicegui-default-gap: 0rem;
        }
     </style>
    ''')

    quote_classes = 'm-3 p-6 bg-blue-100 rounded-lg'
    quote_p_classes = 'font-serif italic text-lg text-gray-700 leading-snug mb-1'
    quote_span_classes = 'text-sm text-gray-500'

    with ui.element('div').classes('flex'):
        with ui.element('div').classes(quote_classes):
            ui.label("The success combination in business is: Do what you do better... and: do more of what you do.").classes(quote_p_classes)
            ui.label("- David Joseph Schwartz").classes(quote_span_classes)

        with ui.element('div').classes(quote_classes):
            ui.label("Give out what you most want to come back.").classes(quote_p_classes)
            ui.label("- Robin Sharma").classes(quote_span_classes)

        with ui.element('div').classes(quote_classes):
            ui.label("The only way around is through.").classes(quote_p_classes)
            ui.label("- Robert Frost").classes(quote_span_classes)

ui.run(title='test example', storage_secret='123', dark=False)
```

I believe that by improving the control over layout and enhancing compatibility with standard HTML/CSS practices, NiceGUI can attract a wider audience of developers. I'm interested in hearing the community's thoughts on this and discussing potential solutions.

**Comment by stepheweffie** [comment]
This is a really simple fix in NiceGUI 
The parent div is no-wrap

```py
with ui.element('div').classes('flex flex-row no-wrap'):
    with ui.element('div').classes(quote_classes):
``` 
<img width="1440" alt="Screen Shot 2024-01-06 at 1 30 15 AM" src="https://github.com/zauberzeug/nicegui/assets/1074503/2f26035b-7528-4ee2-bf1e-7a58d4c1aa8b">

**Reply by Xtreemrus** [reply]
Thank you! :)
In this case, it indeed works, and since the example is quite short, it's easy to pinpoint the issue.

However, the main thought here is that it becomes more challenging to learn Tailwind CSS using NiceGUI, as almost every example won't work exactly the same way.

In fact, there was a similar discussion before, but from the perspective of trying to adapt an entire template:
https://github.com/zauberzeug/nicegui/discussions/1954

NiceGUI has a fantastic architecture, including a fast live reload feature. Perhaps a flag for ui.run could be introduced to make it easier to adapt purchased landing page templates.

**Comment by falkoschindler** [comment]
Thanks for bringing this up, @Xtreemrus!

A major challenge for NiceGUI is the integration of all these powerful frameworks and to please users from different fields. First and foremost we try to build a UI framework that makes it easy to get started. That's why you can simply add some elements to the global scope and you get a page with a flex layout containing vertically stacked UI elements with some padding and some gap in between. That's not how plain HTML or Tailwind or Quasar works. But it feels nice and is sufficient for many use-cases. People coming from [Streamlit](https://streamlit.io/) might wish for even fancier default styling and larger building blocks; others are glad to have control over styling details and access to underlying technologies. Especially web developers are used to HTML, JavaScript, CSS and maybe Tailwind and want to use these frameworks extensively.

Long story short: NiceGUI doesn't behave like an HTML page, like a plain Tailwind app, or exactly like a Quasar app. It consciously deviates here and there to simplify or improve the default behavior or to reduce conflicts between the different worlds involved.

That being said, we're always happy to get feedback and suggestions on how to improve the framework. Discussions like these help to improve it continuously and making the user experience even nicer. 🙂

**Comment by rodja** [comment]
I like the proposal from @Xtreemrus:

> Perhaps a flag for ui.run could be introduced to make it easier to adapt purchased landing page templates.

That way we could keep the defaults "nice and easy" and also please html/css styling professionals (and those who like to become).

**Comment by Xtreemrus** [comment]
Unfortunately, my knowledge isn't sufficient to figure out how to minimize the influence of CSS. I'm just trying to illustrate the point with HTML tags.

HTML code:
```html
<div class="flex flex-col w-full">
  <div class="quote">
    <p>“The success combination in business is: Do what you do better... and: do more of what you do.”</p>
    <span>- David J. Schwartz</span>
  </div>
  <div class="quote">
    <p>“Give out what you most want to come back.”</p>
    <span>- Robin Sharma</span>
  </div>
  <div class="quote">
    <p>“You don't have to be great at something to start, but you have to start to be great at something.”</p>
    <span>- Zig Ziglar</span>
  </div>
</div>
```

Possible NiceGUI code:
```python
from nicegui import ui, Client
from htmltags import Div, Span, Paragraph

@ui.page('/')
async def main(client: Client):
    
    ui.add_head_html('''
     <style>
        :root {
            --nicegui-default-padding: 0rem;
            --nicegui-default-gap: 0rem;
        }
     </style>
    ''')

    quote_classes = 'm-3 p-6 bg-blue-100 rounded-lg'
    quote_p_classes = 'font-serif italic text-lg text-gray-700 leading-snug mb-1'
    quote_span_classes = 'text-sm text-gray-500'

    with Div().classes('flex flex-col w-full'):
        with Div().classes(quote_classes):
            Paragraph("The success combination in business is: Do what you do better... and: do more of what you do.").classes(quote_p_classes)
            Span("- David Joseph Schwartz").classes(quote_span_classes)

        with Div().classes(quote_classes):
            Paragraph("Give out what you most want to come back.").classes(quote_p_classes)
            Span("- Robin Sharma").classes(quote_span_classes)

        with Div().classes(quote_classes):
            Paragraph("The only way around is through.").classes(quote_p_classes)
            Span("- Robert Frost").classes(quote_span_classes)

ui.run(title='test example', storage_secret='123', dark=False)
```

Code for htmltags.py:
```python
from nicegui.elements.mixins.text_element import TextElement as _TextElement
from nicegui.element import Element as _Element

class Div(_Element):

    def __init__(self) -> None:
        """
        The div tag is used in HTML to make divisions of content in the web page (text, images, header, footer, navigation bar, etc).
        """
        super().__init__('div')

class Paragraph(_TextElement):

    def __init__(self, text: str = '') -> None:
        """
        The paragraph (<p> tag in HTML) defines a paragraph.

        :param text: the content of the paragraph
        """
        super().__init__(tag='p', text=text)

class Span(_TextElement):
    
    def __init__(self, text: str = '') -> None:
        """
        The span (<span> tag in HTML) is used to group inline-elements in a document.

        :param text: the content of the span
        """
        super().__init__(tag='span', text=text)
```

Theoretically, the wonderful catalog from the website https://www.geeksforgeeks.org/html-tags-a-to-z-list/ is so well formalized that GPT could generate similar code with descriptions for all elements. For example, the description for the Span class and the class itself were generated by Copilot, just after entering its name, but ideally, for each Tag, only certain attributes are possible, so after defining a suitable class architecture, more complex GPT prompt engineering will likely be required.

A stricter separation into tags (p, h1, span, etc.) may again be necessary to better control the generated HTML code and simplify the porting of CSS and JavaScript.

**Reply by rodja** [reply]
Thanks for the code @Xtreemrus. This is something we are not planning to add to NiceGUI directly because it is quite "low level" and we want NiceGUI to be more about widgets. But this could be a small additional package on top of NiceGUI. Maybe you like to start your own small GitHub project with this?
